lib/pure/ptrarith.nim

```diff
@@ -0,0 +1,25 @@
+#
+#
+#            Nim's Runtime Library
+#        (c) Copyright 2019 Nim contributors
+#
+#    See the file "copying.txt", included in this
+#    distribution, for details about the copyright.
+#
+
+## This module offers operators for performing arithmetic on raw pointer types
+## in Nim. Nim is *not* C: using pointer arithmetic in Nim should be avoided (if
+## possible, use `array[T]`, `ptr UncheckedArray[T]`, etc.), however, pointer
+## arithmetic is sometimes desirable for low-level ops/FFI with C.
+
+func `+`*[T](x: ptr[T]; offset: SomeInteger): ptr[T] =
+  cast[ptr[T]](cast[ByteAddress](x) + cast[ByteAddress](offset * sizeof(T)))
```

Comment:
I think the `+` should be a `+%`, do pointer wraparound like unsigned int? or is it undefined behaviour like signed integers?

Reply:
I assume undefined. But yes I like `+%` more better as a less accessible operator

Reply:
seems better to use `offset` as a function name here - you want to be able to look for where these operations are used instead of hiding them behind an overload

Reply:
see also: https://doc.rust-lang.org/std/primitive.pointer.html - it has docs, examples, risks, etc that hint at how compilers think about this - the names are pretty standard too meaning it's easier to cross-reference with other languages / port code.

Reply:
I actually meant this +% for the ByteAddress summation but +% for the op name sounds good as well: https://nim-lang.org/docs/system.html#%2B%25%2Cint64%2Cint64

I feel like offset is too much. Libraries that import this probably have lots of pointer arithmetics to do.

Reply:
I use ``+!`` for pointer additions, the ``!`` screams "dangerous".

Reply:
`+!` at least is unique and thus possible to grep for. 

I don't really see the argument that you use pointer operations "a lot" - in particular, you typically don't stack multiple pointer ops in a single expression so even with a name like `offset` (that's easier to google for than `%!`), you won't end up with excessively long lines because there will only be one of them

If you have lots and lots of pointer arithmetic to do, perhaps that's a sign that you should wrap it in something safer?

Reply:
That's for writing the wrapper ;)